# TUGAS 6 : PRAKTIKUM ALGORITMA PEMROGRAMAN

NAMA : MOHAMMAD IQBAL SURYA RAMADHAN

NIM  : 210411100002

Algoritma dan pemrograman dasar


1. (Menampilkan piramida)

Tuliskan suatu program yang meminta pengguna memasukkan suatu bilangan bulat
(integer) dari 1 sampai 15 dan tampilkan suatu piramida, sebagaimana diperlihatkan
pada contoh eksekusi di bawah ini:

Masukkan jumlah baris: 7

In [ ]:
n = int(input('Masukkan jumlah baris: '))  

for i in range(n):     
    spaces = ' ' * (2*(n-1-i))     
    print(spaces, end='')
          
    for j in range(i+1, 0, -1):         
        print('%2d' % j, end='')

    for j in range(2, i+2, 1):         
        print('%2d' % j, end='')
             
    print()

             1
           2 1 2
         3 2 1 2 3
       4 3 2 1 2 3 4
     5 4 3 2 1 2 3 4 5
   6 5 4 3 2 1 2 3 4 5 6
 7 6 5 4 3 2 1 2 3 4 5 6 7


2. (Menampilkan bilangan dalam pola piramida)
Tuliskan suatu perulangan for bersarang yang menampilkan output berikut:

In [ ]:
for row in range(0, 9):
    for space in range(8, row, -1):
        print("%5c" %(' '), end=" ")    
    val = 0

    for i in range(0, row):
        print("%5d" %(pow(2, val)), end=" ")
        val += 1
    val1 = row - 1

    for i in range(1, row):
        val1 -= 1
        print("%5d" %(pow(2, val1)), end=" ")
        
    print("")

                                                
                                              1 
                                        1     2     1 
                                  1     2     4     2     1 
                            1     2     4     8     4     2     1 
                      1     2     4     8    16     8     4     2     1 
                1     2     4     8    16    32    16     8     4     2     1 
          1     2     4     8    16    32    64    32    16     8     4     2     1 
    1     2     4     8    16    32    64   128    64    32    16     8     4     2     1 


3. (Menampilkan bilangan prima antara 2 dan 1,000) Perhatikan kode
berikut:

Lakukan perubahan terhadap kode program di atas untuk menampilkan bilangan
prima antara 2 dan 1,000 secara inklusif. Tampilkan delapan bilangan prima per baris.

In [ ]:
NUMBER_OF_PRIMES = 1000 # Number of primes to display
count = 0 # Count the number of prime numbers
number = 2
print("1000 bilangan prima pertama : ")
# Repeatedly find prime numbers
while number <= NUMBER_OF_PRIMES:
  isprime=True
  divisor=2
  
  while divisor < number:
    if (number % divisor == 0):
      isprime=False
      break
    divisor+=1
  
  if (isprime):
    print(format(number,"5d"), end='')
    count+=1
    if count%8==0:
      print("\n")
  number+=1

1000 bilangan prima pertama : 
    2    3    5    7   11   13   17   19

   23   29   31   37   41   43   47   53

   59   61   67   71   73   79   83   89

   97  101  103  107  109  113  127  131

  137  139  149  151  157  163  167  173

  179  181  191  193  197  199  211  223

  227  229  233  239  241  251  257  263

  269  271  277  281  283  293  307  311

  313  317  331  337  347  349  353  359

  367  373  379  383  389  397  401  409

  419  421  431  433  439  443  449  457

  461  463  467  479  487  491  499  503

  509  521  523  541  547  557  563  569

  571  577  587  593  599  601  607  613

  617  619  631  641  643  647  653  659

  661  673  677  683  691  701  709  719

  727  733  739  743  751  757  761  769

  773  787  797  809  811  821  823  827

  829  839  853  857  859  863  877  881

  883  887  907  911  919  929  937  941

  947  953  967  971  977  983  991  997



4. (Menampilkan tahun kabisat)
Tuliskan suatu program yang menampilkan, sepuluh per baris, semua tahun kabisat
dalam abad ke-21 (dari tahun 2001 sampai 2100). Tahun-tahun tersebut dipisahkan
oleh tepat satu spasi.

In [ ]:
year=2001
count=0
count1=0
while year<=2100:
    if year%4==0 and year%100!=0 or year%400==0:
        print(year,end=" ")
        count=count+1
        count1=count1+1
        
        if count==10:
            count=0
            print ()
    year=year+1
print ("\ntotal number of leap year : ", count1)

2004 2008 2012 2016 2020 2024 2028 2032 2036 2040 
2044 2048 2052 2056 2060 2064 2068 2072 2076 2080 
2084 2088 2092 2096 
total number of leap year :  24


5. (Jumlah kemunculan bilangan terbesar)
Tuliskan suatu program yang membaca beberapa integer, mencari nilai terbesar, dan
menghitung jumlah kemunculannya. Anggaplah bahwa input berakhir dengan angka
0. Katakanlah anda memasukkan 3 5 2 5 5 5 0; maka program menemukan bahwa
bilangan terbesar adalah 5 dan jumlah kemunculan untuk 5 adalah 4.
(Hint: Gunakan dua variabel, max dan count. Variabel max menyimpan angka
terbesar terkini, dan count menyimpan jumlah kemunculannya. Pada awalnya,
berikan angka pertama ke max dan 1 untuk count. Bandingkan setiap angka
berikutnya dengan max. Jika angka ini lebih besar daripada max, maka berikan angka
it uke max dan reset count ke 1 lagi. Jika bilangan ini sama dengan max, naikkan count
sebesar 1.)

Masukkan angka (0: untuk mengakhiri): 3
Masukkan angka (0: untuk mengakhiri): 5
Masukkan angka (0: untuk mengakhiri): 2
Masukkan angka (0: untuk mengakhiri): 5
Masukkan angka (0: untuk mengakhiri): 5
Masukkan angka (0: untuk mengakhiri): 5
Masukkan angka (0: untuk mengakhiri): 0
Angka terbesar adalah 5 dengan kemunculan sebanyak 4 kali.

In [ ]:
n = int(input("Masukkan angka (0: untuk mengakhiri): ")) 
largest = n 
count = 1 
while(n!=0):     
    n = int(input("Masukkan angka (0: untuk mengakhiri): "))     
    if(largest == n):         
        count += 1     
    elif(largest < n):         
        largest = n         
        count = 1 
print("Angka terbesar adalah",largest,'dengan kemunculan sebanyak',count,'kali') 

Angka terbesar adalah 5 dengan kemunculan sebanyak 4 kali


6. (Aplikasi keuangan: menghitung nilai investasi masa depan)
Tuliskan suatu fungsi yang menghitung nilai investasi masa depan pada tingkat bunga
tertentu selama jumlah tahun tertentu. Investasi masa depan ditentukan
menggunakan formula:
futureInvestmentValue = investmentAmount * (1 +
monthlyInterestRate)

numberOfMonthsGunakan header fungsi berikut:

def futureInvestmentValue(investmentAmount, monthlyInterestRate, years):
Sebagai contoh, futureInvestmentValue(10000, 0.05/12, 5) mengembalikan 12833.59.
Buatlah suatu program test yang meminta pengguna memasukkan jumlah investasi
(investment amount) dan tingkat bunga tahunan (annual interest rate) dalam persen
dan mencetak tabel yang menampilkan nilai masa depan untuk tahun 1 s.d 30.
Berikut ini adalah contoh hasil eksekusinya:

Jumlah yang diinvestasikan: 1000
Tingkat Bunga Tahunan: 9
Tahun Nilai Masa Depan
1 1093.80
2 1196.41
...
29 13467.25
30 14730.57

In [ ]:
def futureInvestmentValue(investmentAmount, monthlyInterestRate, years):     
    return investmentAmount * (1+monthlyInterestRate)**(years*12)   
    
def main():     
    amount = float(input('Jumlah yang diinvestasikan: '))     
    interest = float(input('Tingkat Bunga Tahunan: '))      
    print('\nTahun\tNilai Masa Depan')     
    for i in range(1, 31):         
        print(str(i) + '\t' + '%.2f' % futureInvestmentValue(amount, interest/1200, i))        
main() 


Tahun	Nilai Masa Depan
1	1093.81
2	1196.41
3	1308.65
4	1431.41
5	1565.68
6	1712.55
7	1873.20
8	2048.92
9	2241.12
10	2451.36
11	2681.31
12	2932.84
13	3207.96
14	3508.89
15	3838.04
16	4198.08
17	4591.89
18	5022.64
19	5493.80
20	6009.15
21	6572.85
22	7189.43
23	7863.85
24	8601.53
25	9408.41
26	10290.99
27	11256.35
28	12312.28
29	13467.25
30	14730.58


7. (Konversi antara kaki dan meter)
Buatlah suatu modul yang mengandung dua fungsi berikut:



In [ ]:
foot = 1
meter = 20

def footToMeter(foot):
    return foot *0.305

def meterToFoot(meter):
    return round((meter/0.305),3)

print('Kaki\tMeter\t|\tMeter\tKaki')
for foot,meter in zip(range(1, 11), range(20, 75, 6)):
    print('%s\t%s\t|\t%s\t%s' % (foot, footToMeter(foot), meter, meterToFoot(meter)))

Kaki	Meter	|	Meter	Kaki
1	0.305	|	20	65.574
2	0.61	|	26	85.246
3	0.915	|	32	104.918
4	1.22	|	38	124.59
5	1.525	|	44	144.262
6	1.83	|	50	163.934
7	2.135	|	56	183.607
8	2.44	|	62	203.279
9	2.745	|	68	222.951
10	3.05	|	74	242.623


8. (Aplikasi keuangan: menghitung komisi)
Buatlah suatu fungsi yang daoat menghitung komisi penjualan menggunakan skema
pada Masalah 9. Header dari fungsi ini adalah

def computeCommission(salesAmount):

Tulislah suatu program test yang menampilkan tabel berikut:
Jumlah Penjualan Komisi

10000            900.0
15000            1500.0
95000            11100.0
100000           11700.0

In [ ]:
#returns the commisson based on the amount

def computeCommission(salesAmount):

    if(salesAmount<=5500):

        return salesAmount*0.08


    return 1500+(salesAmount-15000) * 0.12

#looping thorugh 11000 to 76000

penjualan=10000
print('Jumlah Penjualan Komisi')
while(penjualan<=100000):

    print('%s \t\t %s' % (penjualan,computeCommission(penjualan)))

    penjualan=penjualan+5000

Jumlah Penjualan Komisi
10000 		 900.0
15000 		 1500.0
20000 		 2100.0
25000 		 2700.0
30000 		 3300.0
35000 		 3900.0
40000 		 4500.0
45000 		 5100.0
50000 		 5700.0
55000 		 6300.0
60000 		 6900.0
65000 		 7500.0
70000 		 8100.0
75000 		 8700.0
80000 		 9300.0
85000 		 9900.0
90000 		 10500.0
95000 		 11100.0
100000 		 11700.0


9. (Estimasi )
Estimasi dapat dihitung menggunakan rumus deret berikut:

Tulislah suatu fungsi yang mengembalikan m(i) dari nilai i tertentu dan buatlah suatu
program test yang menampilkan tabel berikut:
i m(i)
1 4.0000
101 3.1515
201 3.1466
301 3.1449
401 3.1441
501 3.1436
601 3.1433
701 3.1430
801 3.1428
901 3.1427

In [ ]:
def calPi(i): 
    mi = 0 
    for j in range(1,i+1):
        mi += 4*((-1)**(j+1))/(2*j-1) 
    return mi

print('i\tm(i)')
for i in range(1,1000,100):
    mi = calPi(i)
    print("{0:d}\t{1:.4f}".format(i, mi))

i	m(i)
1	4.0000
101	3.1515
201	3.1466
301	3.1449
401	3.1441
501	3.1436
601	3.1433
701	3.1430
801	3.1428
901	3.1427


10. (Geometri: menampilkan sudut)

Ubahlah code ComputeDistance.py, menjadi fungsi untuk menghitung jarak antara
dua titik.

contoh inputan : 3,4 dan 6,8

In [ ]:
import math
x1,y1 = eval(input('masukkan x1 dan y1 dari titik 1: '))
x2,y2 = eval(input('masukkan x2 dan y2 dari titik 2: '))
def calculate(x1,y1,x2,y2):
    dist = math.sqrt((x2 - x1)**2 + (y2 - y1)**2)
    return dist 
print(calculate(x1,y1,x2,y2))

5.0
